File: ProblemSolve/290.cpp
/****************
 * Word Pattern *
 ****************/

class Solution {
public:
    vector<string> split(string input, char token){
        vector<string> splitted;
        stringstream ss(input);
        string temp;
        
        while(getline(ss, temp, token)){
            splitted.push_back(temp);
        }
        
        return splitted;
    }
    
    bool wordPattern(string pattern, string s) {
        unordered_map<string, char> hash;
        unordered_map<char, string> hash2;
        vector<string> splitted = split(s, ' ');
        
        if(pattern.size() != splitted.size()){
            return false;
        }
        
        for(int i=0; i<splitted.size(); ++i){
            if(hash.find(splitted[i]) != hash.end() && hash[splitted[i]] != pattern[i]){
                return false;
            }
            
            hash[splitted[i]] = pattern[i];
            
            if(hash2.find(pattern[i]) != hash2.end() && hash2[pattern[i]] != splitted[i]){
                return false;
            }
            
            hash2[pattern[i]] = splitted[i];
        }
        
        return true;
    }
};

File: ProblemSolve/290.cs
/****************
 * Word Pattern *
 ****************/

public class Solution {
    public bool WordPattern(string pattern, string s) {
        Dictionary<char, string> hash = new Dictionary<char, string>();
        Dictionary<string, char> hash2 = new Dictionary<string, char>();
        string[] splitted = s.Split(" ");
        
        if(pattern.Length != splitted.Length){
            return false;
        }
        
        for(int i=0; i<splitted.Length; ++i){
            if(hash.ContainsKey(pattern[i]) == true){
                if(hash[pattern[i]].Equals(splitted[i]) == false){
                    return false;
                }
            }
            else{
                hash.Add(pattern[i], splitted[i]);
            }
            
            if(hash2.ContainsKey(splitted[i]) == true){
                if(hash2[splitted[i]].Equals(pattern[i]) == false){
                    return false;
                }
            }
            else{
                hash2.Add(splitted[i], pattern[i]);
            }
        }
        
        return true;
    }
}